File: 8 String to Integer (atoi)/8string-to-integer-atoi.cpp
class Solution {
private:
    bool is_digit(char c) {
        return c >= '0' && c <= '9';
    }

public:
    int myAtoi(string s) {
        int n = s.size();
        if (n == 0) return 0;

        int i = 0;
        while (i < n and s[i] == ' ') i++;

        if (i == n) return 0;

        // to handle sign cases
        bool is_negative = false;

        if (s[i] == '+') {
            i++;
        }
        else if (s[i] == '-') {
            i++;
            is_negative = true;
        }

        int res = 0;
        while (i < n && is_digit(s[i])) {
            int digit = s[i] - '0';

            // handle overflow res
            if ((res > INT_MAX / 10) || ((res == INT_MAX / 10) && digit > 7)) {
                return is_negative ? INT_MIN : INT_MAX;
            }

            res = res * 10 + digit;
            i++;
        }
        return is_negative ? -res : res;
    }
};

// Time: O(n)
// Space: O(1)

File: 8 String to Integer (atoi)/8string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        # 1. Strip all leading spaces
        i = 0
        while i < len(s) and s[i] == " ":
            i += 1
        if i == len(s):
            return 0

        INT_MAX = 2**31 - 1 # pow(2, 31) - 1
        INT_MIN = -2**31
        
        # 2. Check for the sign
        is_positive = True
        if s[i] == "+":
            i += 1
        elif s[i] == "-":
            i += 1
            is_positive = False
        
        # 3. Parse digits
        res = 0
        while i < len(s) and s[i].isdigit():
            # handle overflow res
            if (res > INT_MAX // 10) or (res == INT_MAX // 10 and int(s[i]) > INT_MAX % 10):
                return INT_MAX if is_positive else INT_MIN

            res = res * 10 + int(s[i])
            i += 1
        return res if is_positive else -res

# Time: O(n)
# Space: O(1)